File: 0086 Partition List/0086.cpp
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        ListNode* lessHead = new ListNode(0);
        ListNode* moreHead = new ListNode(0);
        
        ListNode* less = lessHead;
        ListNode* more = moreHead;
        ListNode* cur = head;

        while (cur){
            if (cur->val < x){
                less->next = cur;
                less = cur;
            } else {
                more->next = cur;
                more = cur;
            }

            cur = cur->next;
        }

        less->next = moreHead->next;
        more->next = nullptr;

        return lessHead->next;
    }
};

File: 0086 Partition List/0086.py
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        less_head, more_head = ListNode(0), ListNode(0)
        less, more = less_head, more_head
        cur = head

        while cur:
            if cur.val < x:
                less.next = cur
                less = less.next
            else:
                more.next = cur
                more = more.next
            cur = cur.next
        
        less.next = more_head.next
        more.next = None

        return less_head.next
